base/compiler/ssair/passes.jl

```diff
@@ -1390,3 +1390,48 @@ function cfg_simplify!(ir::IRCode)
     compact.active_result_bb = length(bb_starts)
     return finish(compact)
 end
+
+function memory_opt!(ir::IRCode, estate)
+    estate = estate::EscapeAnalysis.EscapeState
+    revisit = nothing     # potential targets for a mutating_arrayfreeze drop-in
+    maybecopies = nothing # calls to maybecopy
+
+    # mark statements that possibly can be optimized
+    for idx in 1:length(ir.stmts)
+        stmt = ir.stmts[idx][:inst]
+        isexpr(stmt, :call) || continue
+        if is_known_call(stmt, Core.maybecopy, ir)
+            maybecopies === nothing && (maybecopies = Int[])
+            push!(maybecopies, idx)
+        elseif is_known_call(stmt, Core.arrayfreeze, ir)
+            # array as SSA value might have been initialized within this frame
+            # (thus potentially doesn't escape to anywhere)
+            if isa(stmt.args[2], SSAValue)
+                revisit === nothing && (revisit = Int[])
+                push!(revisit, idx)
+            end
+        end
+    end
+
+    if revisit !== nothing
+        # if array doesn't escape, we can just change the tag and avoid allocation
+        for idx in revisit
+            stmt = ir.stmts[idx][:inst]::Expr
+            arg = stmt.args[2]::SSAValue
+            has_no_escape(estate[arg]) || continue
+            stmt.args[1] = GlobalRef(Core, :mutating_arrayfreeze)
+        end
+    end
+
+    if maybecopies !== nothing
+        for idx in maybecopies
+            println("analyzing maybecopy at ", pc)
```

Comment:
```suggestion
            # println("analyzing maybecopy at ", idx)
```
Please make sure code doesn't yield internal errors.

Reply:
Oops, that’s what I get for pushing changes at the very end of the day 😅